osquery/tables/networking/windows/routes.cpp

```diff
@@ -0,0 +1,158 @@
+
+#include <memory>
+#include <string>
+#include <windows.h>
+#include <winsock2.h>
+
+#include <wS2tcpip.h>
+#include <ws2ipdef.h>
+
+#include <iphlpapi.h>
+#include <mstcpip.h>
+
+#include <boost/algorithm/string/join.hpp>
+#include <osquery/logger.h>
+#include <osquery/tables.h>
+
+#include "osquery/core/conversions.h"
+#include "osquery/core/windows/wmi.h"
+
+namespace osquery {
+namespace tables {
+
+std::map<DWORD, IP_ADAPTER_INFO> getAdapterAddressMapping() {
+  std::map<DWORD, IP_ADAPTER_INFO> returnMapping;
+  auto dwBufLen = 0UL;
+  auto dwStatus = GetAdaptersInfo(NULL, &dwBufLen);
+
+  if (dwStatus != ERROR_BUFFER_OVERFLOW) {
+    return returnMapping;
+  }
+
+  std::vector<BYTE> buffer(dwBufLen);
+  auto pAdapterInfo = reinterpret_cast<PIP_ADAPTER_INFO>(buffer.data());
+  dwStatus = GetAdaptersInfo(pAdapterInfo, &dwBufLen);
+
+  if (dwStatus != NO_ERROR) {
+    return returnMapping;
+  }
+
+  while (pAdapterInfo != nullptr) {
+    auto adapter = *pAdapterInfo;
+    returnMapping.insert(std::make_pair(pAdapterInfo->Index, adapter));
+    pAdapterInfo = pAdapterInfo->Next;
+  }
+
+  return returnMapping;
+}
+
+std::map<unsigned long, MIB_IPINTERFACE_ROW> getInterfaceRowMapping(
+    int type = AF_UNSPEC) {
+  std::map<unsigned long, MIB_IPINTERFACE_ROW> returnMapping;
+  PMIB_IPINTERFACE_TABLE interfaces;
+  auto dwRetVal = GetIpInterfaceTable(type, &interfaces);
+
+  if (dwRetVal != NO_ERROR) {
+    return returnMapping;
+  }
+
+  for (unsigned long i = 0; i < interfaces->NumEntries; ++i) {
+    MIB_IPINTERFACE_ROW currentRow = interfaces->Table[i];
+    returnMapping.insert(std::make_pair(currentRow.InterfaceIndex, currentRow));
+  }
+
+  return returnMapping;
+}
+
+QueryData genRoutes(QueryContext& context) {
+  QueryData results;
+  PMIB_IPFORWARD_TABLE2 ipTable = nullptr;
+  auto result = GetIpForwardTable2(AF_UNSPEC, &ipTable);
+
+  if (result != NO_ERROR) {
+    FreeMibTable(ipTable);
+
+    return results;
+  }
+
+  auto numEntries = ipTable[0].NumEntries;
+  auto interfaces = getInterfaceRowMapping();
+  auto adapters = getAdapterAddressMapping();
+
+  for (unsigned long i = 0; i < numEntries; ++i) {
+    Row r;
+    std::string interfaceIpAddress;
+    const auto& currentRow = ipTable[0].Table[i];
+    auto addrFamily = currentRow.DestinationPrefix.Prefix.si_family;
+    auto actualInterface = interfaces.at(currentRow.InterfaceIndex);
+    if (addrFamily == AF_INET6) {
+      std::vector<char> buffer(INET6_ADDRSTRLEN);
+
+      r["mtu"] = INTEGER(actualInterface.NlMtu);
+      // These are all technically "on-link" addresses according to
+      // `route print -6`.
+      r["type"] = "local";
+      auto ipAddress = currentRow.DestinationPrefix.Prefix.Ipv6.sin6_addr;
+      auto gateway = currentRow.NextHop.Ipv6.sin6_addr;
+
+      InetNtop(addrFamily, (PVOID)&ipAddress, buffer.data(), buffer.size());
+      r["destination"] = SQL_TEXT(buffer.data());
+      InetNtop(addrFamily, (PVOID)&gateway, buffer.data(), buffer.size());
```

Comment:
Would it be worth doing a 'buffer.clear()' before using the buffer again? seems like the gateway address will fully overwrite the IP though..